ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+    result = PaymentService.execute(user.payment_method)
+    raise PaymentFailed, "Unable to process payment: #{result.error}" if result.failure?
+
+    result.payment_id
+  end
+end
+```
+
+Notes:
+
+- Errors defined within the activity class are namespaced as `MyActivity::PaymentFailed`,
+  which is a very useful property
+  - This will however not work if a workflow does not have access to the activity definition
+  - In this case it will get raised as a generic `Temporal::Error::ActivityFailure` with an original
+    error's class name available via `#original_error`
+- Subclassing `Temporal::Error::ActivityFailure` allows us to inject extra information into the
+  error (such as activity name, id, etc)
+- All errors will also have their backtrace serialized and relayed to the caller
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
+
+The main difference is that a DataConverter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the later is for
+modifying a binary that is transmitted over the wire.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7933')
+
+config.add_data_converter(MyProtobufConverter.new)
+config.add_data_converter(MyThriftConverter) # can be a class/module based on duck-typing
+
+config.add_codec(EncryptionCodec.new(KEY_ID))
+config.add_codec(CompressionCodec)
+```
+
+A data converter is expected to have this interface:
+
+```ruby
+class Temporal::DataConverter::Base
+  # Returns mime-type of the payload
+  def encoding() -> String
+
+  # Takes any Ruby object and returns a Payload object or nil if unable to convert
+  def to_payload(data: Any) -> Temporal::Payload
+
+  # Takes a Payload object and returns a Ruby object. Only called on a matching encoding
+  def from_payload(payload: Temporal::Payload) -> Any
+end
+```
+
+And a code is expected to have this interface:
+
+```ruby
+class Temporal::Codec::Base
+  # Takes Payloads object and returns a modified Payloads object
+  def encode(payloads: [Temporal::Payload]) -> [Temporal::Payload]
+
+  # Takes Payloads object and returns a modified Payloads object
+  def decode(payloads: [Temporal::Payload]) -> [Temporal::Payload]
+end
+```
+
+### Alternatives considered
+
+While some other SDKs combine these two constructs into one with a broader interface we think that
+these are different enough concepts that deserve first-class separation. Otherwise implementing
+Codecs becomes a tricky task where each Codec needs to be aware of all the DataConverters available.
```

Comment:
Concur, though based on my comment above, I think what you call a data converter is what others call a payload converter.

Reply:
Renamed here — https://github.com/temporalio/proposals/pull/57/commits/804a864ebb82ec3015836942b09ea6411ad0efcd